pkg/hub/managedclusterset/controller_test.go

```diff
@@ -0,0 +1,349 @@
+				}
+
+				clusterSet = actions[1].(clienttesting.UpdateAction).GetObject().(*clusterv1alpha1.ManagedClusterSet)
+				if !hasCondition(
+					clusterSet.Status.Conditions,
+					clusterv1alpha1.ManagedClusterSetConditionEmpty,
+					metav1.ConditionFalse,
+					"ClustersSelected",
+					"1 ManagedClusters selected") {
+					t.Errorf("expected conditon is not found: %v", clusterSet.Status.Conditions)
+				}
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			objects := []runtime.Object{}
+			for _, cluster := range c.existingClusters {
+				objects = append(objects, cluster)
+			}
+			for _, clusterSet := range c.existingClusterSets {
+				objects = append(objects, clusterSet)
+			}
+
+			clusterClient := clusterfake.NewSimpleClientset(objects...)
+
+			informerFactory := clusterinformers.NewSharedInformerFactory(clusterClient, 5*time.Minute)
+			for _, cluster := range c.existingClusters {
+				informerFactory.Cluster().V1().ManagedClusters().Informer().GetStore().Add(cluster)
+			}
+			for _, clusterSet := range c.existingClusterSets {
+				informerFactory.Cluster().V1alpha1().ManagedClusterSets().Informer().GetStore().Add(clusterSet)
+			}
+
+			ctrl := managedClusterSetController{
+				clusterClient,
+				informerFactory.Cluster().V1().ManagedClusters().Lister(),
+				informerFactory.Cluster().V1alpha1().ManagedClusterSets().Lister(),
+				eventstesting.NewTestingEventRecorder(t)}
+
+			syncErr := ctrl.sync(context.TODO(), testinghelpers.NewFakeSyncContext(t, c.queueKey))
+			if syncErr != nil {
+				t.Errorf("unexpected err: %v", syncErr)
+			}
+
+			c.validateActions(t, clusterClient.Actions())
+		})
+	}
+}
+
+func newManagedCluster(name string, labels map[string]string) *clusterv1.ManagedCluster {
+	return &clusterv1.ManagedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: labels,
+		},
+	}
+}
+
+func newManagedClusterSet(name string, clusterSelectors []clusterv1alpha1.ClusterSelector, terminating bool) *clusterv1alpha1.ManagedClusterSet {
+	clusterSet := &clusterv1alpha1.ManagedClusterSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: clusterv1alpha1.ManagedClusterSetSpec{
+			ClusterSelectors: clusterSelectors,
+		},
+	}
+	if terminating {
+		now := metav1.Now()
+		clusterSet.DeletionTimestamp = &now
+	}
+
+	return clusterSet
+}
+
+func hasCondition(conditions []clusterv1alpha1.StatusCondition, conditionType string, status metav1.ConditionStatus, reason, message string) bool {
+	for _, condition := range conditions {
+		if condition.Type != conditionType {
+			continue
+		}
+		if condition.Status != status {
+			continue
+		}
+		if condition.Reason != reason {
+			continue
+		}
+		if condition.Message != message {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+// elementsMatch compares x is equal to y ignoring the order of the elements.
+// It assumes that there is no duplicated item in the input slices
+func elementsMatch(xs, ys []string) bool {
```

Comment:
You'll be able to take this out once you use stringset.

Reply:
Code refactored with sets.String.